Update Makefiles

clumens opened the pull request.

kgaillot commented on GNUmakefile:
```diff
@@ -59,3 +58,5 @@ clang $(COVERAGE_TARGETS) $(COVERITY_TARGETS) cppcheck indent:
 		CPPCHECK_ARGS=$(CPPCHECK_ARGS)		\
 		-C devel "$@"
+
+FORCE:
```
Should FORCE be phony?

clumens replied:
I don't think it needs to be, but the docs don't really say.  What they do say is that the prerequisites of PHONY can't be patterns because '%' is always interpreted literally, and that because the FORCE target has no prerequisites or rules, `make` will assume it has always been updated which will force anything depending on it to be rebuilt.

kgaillot commented on cts/Makefile.am:
```diff
@@ -23,28 +23,7 @@ dist_test_DATA		= README.md			\
-		  cli/regression.upgrade.exp			\
-		  cli/regression.validity.exp			\
-		  cli/regression.access_render.exp
+dist_cli_DATA	= $(wildcard cli/*.xml) $(wildcard cli/*.exp)
```
I believe you can list multiple patterns in one wildcard call

clumens replied:
Can and did.  Fixed.

kgaillot commented on xml/version-diff.sh:
```diff
@@ -0,0 +1,52 @@
+#!/bin/sh
```
add a copyright notice

clumens replied:
Added, with the starting copyright date being the year the stuff was added to xml/Makefile.am in the first place.

kgaillot commented on cts/cts.in:
```diff
@@ -156,15 +156,21 @@ if [ x$cmd != x ]; then
-	    cluster-helper -g $cluster_name -- sed -i "s/.*PCMK_trace_functions=/PCMK_trace_functions=$trace,/" $config
+	    cp "$config" "$config.$$"
+	    cluster-helper -g $cluster_name -- sed -e "s/.*PCMK_trace_functions=/PCMK_trace_functions=$trace,/" "$config" > "$config.$$"
+	    mv -- "$config.$$" "$config"
```
The changes in this file won't work -- cluster-helper logs into each node and runs the given command. Not sure whether the `\&\&` trick will work with cluster-helper.

I think the point of calling `cp` first is to use `-p` so the new file has the same permissions as the old. Otherwise it's redundant.

Might be worth functionizing if argument quoting isn't too much of a pain

clumens replied:
Ah right, yeah - I think I realized that about cluster-init.in and then didn't go back and fix cts.in.  I'll change that.  The `\&\&` thing should work.

clumens replied:
The only problem with functionizing it is I don't think we have any shell library I can source from all these places, do we?  It would have to be accessible from cts/, maint/, and tools/.

kgaillot replied:
I wasn't thinking of a library, just duplicating it where needed

kgaillot commented on cts/cts.in:
```diff
@@ -156,15 +156,21 @@ if [ x$cmd != x ]; then
 	    echo "Removing $trace from PCMK_trace_functions"
-	    cluster-helper -g $cluster_name -- sed -i "s/.*PCMK_trace_functions=\\\\\\(.*\\\\\\)$trace,\\\\\\(.*\\\\\\)/PCMK_trace_functions=\\\\\\1\\\\\\2/" $config
+	    cp "$config" "$config.$$"
+	    cluster-helper -g $cluster_name -- sed -e "s/.*PCMK_trace_functions=\\\\\\(.*\\\\\\)$trace,\\\\\\(.*\\\\\\)/PCMK_trace_functions=\\\\\\1\\\\\\2/" "$config" > "$config.$$"
```
what the heck is this pattern doing ...

clumens replied:
If I were to guess, I would say it's trying to remove whatever functions are in $trace from the PCMK_trace_functions= list.  But because it's in one shell script, that gets passed as an argument to another shell script, that gets passed through ssh to a shell on a remote host, to sed, everything has to be escaped a thousand times.

I guess my first question would be:  is anyone actually using this?  And then if so, maybe we should break this out into a tiny script that gets installed on the remote system, and that gets called and we can skip all this ridiculous escaping.

kgaillot replied:
I vaguely recall seeing some use of the cluster-* scripts reported upstream. Probably worth keeping around for now. If the pattern works, I'd say just leave it alone

kgaillot commented on lib/cib/Makefile.am:
```diff
@@ -12,8 +12,13 @@ include $(top_srcdir)/mk/common.mk
+			  cib_native.c 	\
+			  cib_attrs.c
+libcib_la_SOURCES	+= cib_file.c 	\
+			   cib_remote.c
```
No need for the +=, it was just a way to break it up on two lines

clumens replied:
Fixed.

kgaillot replied:
?

kgaillot reviewed (commented):
whew, that needed more than I expected

clumens commented on xml/version-diff.sh:
```diff
@@ -0,0 +1,52 @@
+
+                    echo "$p" | sed -n -e "$((hline - 2)),$hline!d" -e '/^\(+++\|---\)/p'
+                    echo "$h context: $ctxt"
+                    echo "$p" | sed -n -e "1,${hline}d" -e '/^\(---\|@@ \)/be;p;d;:e;n;be'
```
An awful lot of this could use some comments, too.  What do all those sed commands even mean?

clumens commented:
Updated to address review and added a few comments on top to fix bugs in cts/cts.in that I found while testing it for the sed changes.  Note that I have not tested abi-check, bumplibs, or cluster-init but they look good.

kgaillot commented on cts/cts.in:
```diff
@@ -161,19 +161,36 @@ if [ x$cmd != x ]; then
-        sed_in_place_remotely "$config" "s/.*PCMK_trace_functions=/PCMK_trace_functions=$trace,/"
+        # Note that this only works if there's already a PCMK_trace_functions line.
+        # If there isn't one, create it with trace-set first.
+        sed_in_place_remotely "$config" "'/^PCMK_trace_functions=/ s/\"$/,$trace\"/'"
```
technically the quotes are optional in sysconfig, and there could be whitespace before the variable name or between last quote and end of line (or even a # comment at the end). we don't have to handle every possible combination, but maybe whatever can be done in one command

clumens replied:
Fixed with an awful lot of regex work, which is hopefully commented well enough for whoever looks at this next.

kgaillot reviewed (commented):
whew

clumens commented on cts/cts.in:
```diff
@@ -156,15 +160,15 @@ if [ x$cmd != x ]; then
 	trace-add|tadd)
 	    echo "Adding $trace to PCMK_trace_functions"
-	    cluster-helper -g $cluster_name -- sed -i "s/.*PCMK_trace_functions=/PCMK_trace_functions=$trace,/" $config
+        sed_in_place_remotely "$config" "s/.*PCMK_trace_functions=/PCMK_trace_functions=$trace,/"
```
Added a patch on top to get rid of the tab characters and make the spacing and indentation consistent throughout.

clumens commented on tools/cluster-init.in:
```diff
@@ -84,6 +84,16 @@ function helptext() {
+function sed_in_place_remotely() {
+    ssh -l root $1 -- cp $2 $2.sed \&\& sed -e "$3" $2 \> $2.sed \&\& mv $2.sed $2
+}
+
```
This will cascade up to getting rid of the --setup argument to cts/cts, which obviously isn't a big deal, but makes me start wondering how useful the clean and install arguments are as well.

kgaillot replied:
clean/install should still be usable

clumens commented on xml/version-diff.sh:
```diff
@@ -0,0 +1,60 @@
+                    iline=$(grep -Fn "$i" "$v" | cut -d: -f1)
+
+                    if [ "$(echo "$iline" | wc -l)" = "1" ]; then
+                        ctxt=$({ sed -n -e "1,$((iline - 1))p" "$v"
```
Fixed.  Note that if I'm reading this correctly, `make` was using SHELL=/bin/sh before.

clumens commented on cts/cts.in:
```diff
@@ -33,6 +33,10 @@ trace=""
 patterns="-e CTS:"
 
+function sed_in_place_remotely() {
+    cluster-helper -g $cluster_name -- cp $1 $1.sed \&\& sed -e "$2" "$1" \> "$1.sed" \&\& mv $1.sed $1
```
Fixed.
